src/cluster_legacy.c

```diff
@@ -1192,6 +1192,35 @@ void clusterInitLast(void) {
 
 /* Called when a cluster node receives SHUTDOWN. */
 void clusterHandleServerShutdown(void) {
+    if (server.auto_failover_on_shutdown) {
+        /* Find the first best replica, that is, the replica with the largest offset. */
+        client *best_replica = NULL;
+        listIter replicas_iter;
+        listNode *replicas_list_node;
+        listRewind(server.replicas, &replicas_iter);
+        while ((replicas_list_node = listNext(&replicas_iter)) != NULL) {
+            client *replica = listNodeValue(replicas_list_node);
+            /* This is done only when the replica offset is caught up, to avoid data loss */
+            if (replica->repl_state == REPLICA_STATE_ONLINE && replica->repl_ack_off == server.primary_repl_offset) {
+                best_replica = replica;
+                break;
+            }
+        }
+
+        if (best_replica) {
+            /* Send a CLUSTER FAILOVER FORCE to the best replica. */
+            const char *buf = "*3\r\n$7\r\nCLUSTER\r\n$8\r\nFAILOVER\r\n$5\r\nFORCE\r\n";
+            if (connWrite(best_replica->conn, buf, strlen(buf)) == (int)strlen(buf)) {
```

Comment:
ok, i think here we have a problem, the primary is sending a CLUSTER FAILOVER to replica, replica will feed it to the replication stream.
```
    /* If the client is a primary we need to compute the difference
     * between the applied offset before and after processing the buffer,
     * to understand how much of the replication stream was actually
     * applied to the primary state: this quantity, and its corresponding
     * part of the replication stream, will be propagated to the
     * sub-replicas and to the replication backlog. */
    if (c->flag.primary) {
        long long applied = c->reploff - prev_offset;
        if (applied) {
            replicationFeedStreamFromPrimaryStream(c->querybuf + c->repl_applied, applied);
            c->repl_applied += applied;
        }
    }
```

I always though a ADMIN command won't do this thing, look like i get it wrong all the time. Do you guys think there is an another way to do this?

Reply:
What if it needs auth? A replica uses `primaryuser` and `primaryauth` to connect to the primary. At any time, failover can happen to change the roles of primary and replica. Therefore, I guess it's OK that the primary can use `primaryuser` and `primaryauth` to connect to the replica too...?

Reply:
Maybe we can implement a FAILOVER command in cluster mode? see https://github.com/valkey-io/valkey/commit/0d18a1e85fdb9967419dc9f7a5905452d065c84a
```
int clusterAllowFailoverCmd(client *c) {
    if (!server.cluster_enabled) {
        return 1;
    }
    addReplyError(c, "FAILOVER not allowed in cluster mode. "
                     "Use CLUSTER FAILOVER command instead.");
    return 0;
}
```

After the FAILOVER command is used, the primary node will psync with the replica, it is not a primary client, so it will not be put into the replication stream. Maybe we can implement it like this. However, replication.c and cluster.c are separate, which is a bit troublesome.

Reply:
You mean the primary sends PSYNC FAILOVER to the replica? 🤔 We add some code for cluster, like `if (cluster) start failover vote`?

If we send this to an old version node, it can be very confused? If it becomes a primary but it does not do a cluster vote, we hvae a very confused cluster?